plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetescluster/KubernetesClusterManagerImpl.java

```diff
@@ -0,0 +1,3061 @@
+            throw new ManagementServerException(msg);
+        }
+
+        int sshPort = publicIpSshPort.second();
+        KubernetesClusterDetailsVO kubeConfigDetail = kubernetesClusterDetailsDao.findDetail(kubernetesCluster.getId(), "kubeConfigData");
+        if (kubeConfigDetail == null || Strings.isNullOrEmpty(kubeConfigDetail.getValue())) {
+            boolean k8sKubeConfigCopied = false;
+            String kubeConfig = getKubernetesClusterConfig(kubernetesCluster, publicIpAddress, sshPort, 5);
+            if (!Strings.isNullOrEmpty(kubeConfig)) {
+                k8sKubeConfigCopied = true;
+            }
+            if (!k8sKubeConfigCopied) {
+                String msg = String.format("Failed to start Kubernetes cluster ID: %s in usable state as unable to retrieve kube-config for the cluster", kubernetesCluster.getUuid());
+                LOGGER.error(msg);
+                stateTransitTo(kubernetesCluster.getId(), KubernetesCluster.Event.OperationFailed);
+                throw new ServerApiException(ApiErrorCode.INTERNAL_ERROR, msg);
+            }
+            kubernetesClusterDetailsDao.addDetail(kubernetesCluster.getId(), "kubeConfigData", Base64.encodeBase64String(kubeConfig.getBytes(Charset.forName("UTF-8"))), false);
+        }
+        KubernetesClusterDetailsVO dashboardServiceRunningDetail = kubernetesClusterDetailsDao.findDetail(kubernetesCluster.getId(), "dashboardServiceRunning");
+        if (kubeConfigDetail == null || !Boolean.parseBoolean(dashboardServiceRunningDetail.getValue())) {
+            boolean dashboardServiceRunning = isKubernetesClusterDashboardServiceRunning(kubernetesCluster, publicIpAddress, sshPort, 10, 20000);
+            if (!dashboardServiceRunning) {
+                String msg = String.format("Failed to start Kubernetes cluster ID: %s in usable state as unable to get Dashboard service running for the cluster", kubernetesCluster.getUuid());
+                LOGGER.error(msg);
+                stateTransitTo(kubernetesCluster.getId(), KubernetesCluster.Event.OperationFailed);
+                throw new ServerApiException(ApiErrorCode.INTERNAL_ERROR, msg);
+            }
+            kubernetesClusterDetailsDao.addDetail(kubernetesCluster.getId(), "dashboardServiceRunning", String.valueOf(dashboardServiceRunning), false);
+        }
+
+        stateTransitTo(kubernetesCluster.getId(), KubernetesCluster.Event.OperationSucceeded);
+        LOGGER.debug(String.format("Kubernetes cluster ID: %s successfully started", kubernetesCluster.getUuid()));
+        return true;
+    }
+
+    // Open up  firewall port CLUSTER_API_PORT, secure port on which Kubernetes API server is running. Also create port-forwarding
+    // rule to forward public IP traffic to master VM private IP
+    // Open up  firewall ports NODES_DEFAULT_START_SSH_PORT to NODES_DEFAULT_START_SSH_PORT+n for SSH access. Also create port-forwarding
+    // rule to forward public IP traffic to all node VM private IP
```

Comment:
comments like these should end up being method calls inside the method to clean up the code.
```
openUpFireWall(..);
secureKubernetesApiPort(..);
createPortForwardingRules(..);
...
```
this method is way too long